src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,158 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        T oldValue = foundNode.value;
+        foundNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        T value = foundNode.value;
+        unlink(foundNode);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node foundNode = getElementByValue(object);
+        if (foundNode == null) {
+            return false;
+        }
+        unlink(foundNode);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is out of the bounds");
+        }
+    }
+
+    private Node getElementByIndex(int index) {
+        boolean descendingOrder = index > (size / 2);
+
+        Node node = descendingOrder ? tail : head;
+        int currentIndex = descendingOrder ? size - 1 : 0;
+        while (node != null) {
+            if (index == currentIndex) {
+                return node;
+            }
+            node = descendingOrder ? node.prev : node.next;
+            currentIndex = descendingOrder ? currentIndex - 1 : currentIndex + 1;
+        }
+        throw new NoSuchElementException("There is no such element");
+    }
+
+    private Node getElementByValue(T value) {
+        Node node = head;
+        while (node != null) {
+            if (valuesAreEqual(node.value, value)) {
+                break;
+            }
+            node = node.next;
+        }
+        return node;
+    }
+
+    private boolean valuesAreEqual(T firstValue, T secondValue) {
+        return firstValue == secondValue || (firstValue != null && firstValue.equals(secondValue));
+    }
+
+    private void link(Node prev, T value, Node next) {
+        Node newNode = new Node(prev, next, value);
+
+        if (next == null) {
+            tail = newNode;
+        } else {
+            next.prev = newNode;
+        }
+
+        if (prev == null) {
+            head = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
+    }
+
+    private void unlink(Node foundNode) {
+        if (foundNode.prev == null) {
+            head = foundNode.next;
+        } else {
+            foundNode.prev.next = foundNode.next;
+        }
+
+        if (foundNode.next == null) {
+            tail = foundNode.prev;
+        } else {
+            foundNode.next.prev = foundNode.prev;
+        }
+        size--;
+    }
+
+    private class Node {
```

Comment:
Make it not dependant of the outer class
```suggestion
    private static class Node<T> {
```